File: zeroinstall/cmd/add.py
"""
The B{0install add} command-line interface.
"""

# Copyright (C) 2012, Thomas Leonard
# See the README file for details, or visit http://0install.net.

from __future__ import print_function

import sys

from zeroinstall import _
from zeroinstall.cmd import UsageError, select
from zeroinstall.injector import model, requirements

syntax = "PET-NAME INTERFACE"

def add_options(parser):
	select.add_options(parser)

def handle(config, options, args):
	"""@type args: [str]"""
	if len(args) != 2:
		raise UsageError()

	pet_name = args[0]
	iface_uri = model.canonical_iface_uri(args[1])

	sels = select.get_selections(config, options, iface_uri, select_only = False, download_only = True, test_callback = None)
	if not sels:
		sys.exit(1)	# Aborted by user

	root_feed = config.iface_cache.get_feed(iface_uri)
	if root_feed:
		target = root_feed.get_replaced_by()
		if target is not None:
			print(_("Warning: interface {old} has been replaced by {new}".format(old = iface_uri, new = target)))

	r = requirements.Requirements(iface_uri)
	r.parse_options(options)

	app = config.app_mgr.create_app(pet_name, r)
	app.set_selections(sels)
	app.integrate_shell(pet_name)

def complete(completion, args, cword):
	"""@type completion: L{zeroinstall.cmd._Completion}
	@type args: [str]
	@type cword: int"""
	if cword != 1: return
	completion.expand_interfaces()

File: ocaml/cli/add.ml
(* Copyright (C) 2013, Thomas Leonard
 * See the README file for details, or visit http://0install.net.
 *)

(** The "0install add" command *)

open Options
open Support
open Support.Common

module Apps = Zeroinstall.Apps
module U = Support.Utils
module R = Zeroinstall.Requirements
module F = Zeroinstall.Feed

(** Warn the user if [uri] has been replaced. *)
let check_for_replacement f config uri =
  let feed = Zeroinstall.Feed_url.master_feed_of_iface uri in
  match Zeroinstall.Feed_cache.get_cached_feed config feed with
  | None -> log_warning "Master feed for '%s' missing!" uri
  | Some feed ->
      match F.replacement feed with
      | Some replacement ->
        Format.fprintf f "Warning: interface %s has been replaced by %s@." uri replacement
      | None -> ()

let handle options flags args =
  let select_opts = ref [] in
  let refresh = ref false in
  Support.Argparse.iter_options flags (function
    | #common_option as o -> Common_options.process_common_option options o
    | #binary_select_option as o -> select_opts := o :: !select_opts
    | `Refresh -> refresh := true
  );
  match args with
  | [pet_name; arg] -> (
      let module G = Generic_select in
      match G.resolve_target options.config !select_opts arg with
      | (G.Interface, reqs) -> (
          match G.get_selections options ~refresh:!refresh reqs `Download_only |> Lwt_main.run with
          | `Aborted_by_user -> raise (System_exit 1)
          | `Success sels ->
              check_for_replacement options.stdout options.config reqs.R.interface_uri;
              let app = Apps.create_app options.config pet_name reqs in
              Apps.set_selections options.config app sels ~touch_last_checked:true;
              Apps.integrate_shell options.config app pet_name
      )
      | (G.Selections _, _) -> Safe_exn.failf "'%s' is a selections document, not an interface URI" arg
      | (G.App _, _) -> Safe_exn.failf "'%s' is an app, not an interface URI" arg
  )
  | _ -> raise (Support.Argparse.Usage_error 1)
